EZP-29714: As an Editor, I want to use pagination in the sub-items widgets

tischsoic opened the pull request:
# Pagination for Sub-items
Implementation of pagination for Sub-items component.

## Ticket
JIRA: https://jira.ez.no/browse/EZP-29714

## Edge cases for testing (for me and *QA*)
- when you delete/move last item from the given page it should fall back to previous page (if exists)
- when you delete/move all sub-item it should display information that there are no items
- if elements are sorted by *Priority* and you change priority of some element, all pages data should be discarded and loaded again
- if there are no sub-items switch view button should be disabled

## Screenshots

![screen shot 2018-10-31 at 16 11 06](https://user-images.githubusercontent.com/10233057/47797939-a49cc280-dd27-11e8-9c60-a779fed2adfa.png)

When there are no subitems: 
![screen shot 2018-11-06 at 14 54 44](https://user-images.githubusercontent.com/10233057/48070522-53357d00-e1d8-11e8-9b1a-6970cb6a2bdc.png)

Pagination view with many pages:
![screen shot 2018-11-07 at 11 35 51](https://user-images.githubusercontent.com/10233057/48126542-b0d1d400-e281-11e8-8077-c1226ef212e1.png)

tischsoic committed:
EZP-29714 - Implement pagination in Sub-items

tischsoic commented:
Tests are currently failing because we replaced "Load more" with pagination hence the behaviour has changed. Tests need to be adjusted to new pagination.

tischsoic commented:
ping @sunpietro @dew326 - frontend is ready for review

tischsoic commented:
ping @sunpietro @dew326 - pagination algorithm changed

dew326 commented on src/modules/sub-items/components/pagination/pagination.btn.component.js:
```diff
@@ -0,0 +1,38 @@
+
+    return (
+        <li className={className}>
+            <button className="page-link" onClick={onClick}>
```
`onClick` -> `onClick` is for me not proper. Because `on click` you do not do `on click` but rather `handle click`

tischsoic replied:
I like this proposition. I was curious to know how it was for the rest of the code in `admin-ui-modules`, so I made small search. For `={han` search there are 2 results, and for `={on` there are 25 results, which have been written by me, you and @sunpietro . 😛 

I will gladly change it to `handleClick` because it feels better but it's hard to write code which satisfies others when there are no clear rules written down. I think I will start writing them down. 😄

lserwatka replied:
@tischsoic It is a very good idea!

dew326 replied:
There is so many `={on` because of passing prop from parent to the child component. This is ok, the prop can be named `onSomething`, we should avoid naming method `onSomething`. 

When I checked for `={this.on` there are a few matches (mostly from the beginning of v2 where we were in kinda rush).

tischsoic replied:
Thank you for reply and clarification @dew326 , now it makes sense to me. :)

tischsoic replied:
During implementation of bulk delete/move I introduced `onSelectAll` method: https://github.com/ezsystems/ezplatform-admin-ui-modules/blob/946fdedb0895ffb7ed811adcfb0884bdb64adbbc/src/modules/sub-items/components/table-view/table.view.component.js#L154
I'm wondering, should it be named `selectAll` or `handleSelectAll`? ping @dew326 :)

tischsoic replied:
OK, now I saw other comments and I understand it should be renamed to `selectAll`.

dew326 commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,129 @@
+    renderPrevious() {
+        const { currentPage, onPageChange } = this.props;
+        const previousPage = currentPage - 1;
+        const hasPreviousPage = currentPage !== 1;
```
Maybe just `isDisabled = currentPage === 1`? or you can name it `isFirstPage`.

dew326 commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,129 @@
+        return (
+            <PaginationBtnComponent
+                page={previousPage}
+                label="Back"
```
translation

dew326 commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,129 @@
+    renderNext() {
+        const { currentPage, onPageChange } = this.props;
+        const nextPage = currentPage + 1;
+        const hasNextPage = currentPage !== this.getPagesNumber();
```
The same story like in `hasPreviousPage`. `isDisabled` or `isLastPage`, `currentPage === this.getPagesNumber()`

dew326 commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,129 @@
+        return (
+            <PaginationBtnComponent
+                page={nextPage}
+                label="Next"
```
translation

dew326 commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,129 @@
+
+        return pages.map((page) => {
+            if (page === '...') {
+                return <PaginationBtnComponent label="..." disabled={true} />;
```
This component also should have some `key` prop.

tischsoic replied:
Nice catch, interesting that I don't recall any React warnings. 🤔 But it should have a key.

dew326 commented on src/modules/sub-items/components/table-view/table.view.component.js:
```diff
@@ -249,16 +155,16 @@ export default class TableViewComponent extends Component {
                         <input type="checkbox" checked={anyLocationSelected} onChange={this.onSelectAll} />
                     </td>
-                    <td className={`${cellHeadClass} ${cellClass}--name ${cellSortClass}`} onClick={this.sortByName}>
+                    <td className={`${cellHeadClass} ${cellClass}--name ${cellSortClass}`} onClick={this.onSortByName}>
```
The previous one was proper. `onClick` do `sortByName` not `onSortByName`.
Also applies to `onSortByDate` and `onSortByPriority`

dew326 commented on src/modules/sub-items/components/viewing-message/viewing.message.component.js:
```diff
@@ -0,0 +1,30 @@
+
+    return (
+        <div className="c-viewing-message">
+            <div className="c-viewing-message__message" dangerouslySetInnerHTML={{ __html: message }} />
```
`c-viewing-message__message` looks not so good. Maybe `c-viewing-message__text`?

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -605,18 +806,65 @@ export default class SubItemsModule extends Component {
+                    onItemSelect={this.toggleItemSelect}
+                    toggleAllItemsSelect={this.toggleAllPageItemsSelection}
+                    selectedLocationsIds={selectedPageLocationsIds}
+                    onSortChange={this.onSortChange}
```
`onSortChange ={this.changeSorting}`.

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -572,23 +750,46 @@ export default class SubItemsModule extends Component {
+                itemsPerPage={itemsPerPage}
+                currentPage={activePage + 1}
+                totalCount={totalCount}
+                onPageChange={this.onPageChange}
```
`onPageChange={this.changePage}`.

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -60,9 +73,26 @@ export default class SubItemsModule extends Component {
+        const { activePage, pages, totalCount} = this.state;
+        const { limit: itemsPerPage } = this.props;
+        const numberOfPages = Math.ceil(totalCount / itemsPerPage);
+        const pageDoesNotExist = activePage > numberOfPages - 1 && activePage !== 0;
```
`isPageExisting`?

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -74,14 +104,34 @@ export default class SubItemsModule extends Component {
+    }
+
+    updateListViewHeight() {
+        const listViewNode = document.querySelector('.m-sub-items__list');
```
Maybe `ref`?

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -90,32 +140,34 @@ export default class SubItemsModule extends Component {
-                window.eZ.helpers.notification.showErrorNotification('An error occurred while loading items in the Sub Items module')
-            );
+            .catch(() => {
+                window.eZ.helpers.notification.showErrorNotification('An error occurred while loading items in the Sub Items module');
```
Could you add translation?

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -196,22 +252,26 @@ export default class SubItemsModule extends Component {
+        const pageItems = locations.map((location) => {
             const itemLocation = location.value.Location;
+            const respectiveContent = contents.find((item) => item.value.Content._id === itemLocation.ContentInfo.Content._id).value
+                .Content;
```
When this `.Content` is in another line it looks weird. Can we do something with this?

tischsoic replied:
I solved it by adding `map` in `loadContentItems`: 
`contents: contentInfo.View.Result.searchHits.searchHit.map((searchHit) => searchHit.value.Content)`.
Now we don't have `.Content` in another line and `contents` contains `Content`s and not `searchHit.value.Content`s.

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +300,49 @@ export default class SubItemsModule extends Component {
     }
 
+    updateTotalCountState(totalCount) {
+        this.setState(() => {
```
Is this working without the brackets? Shouldn't be `({ totalCount })`?

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +300,49 @@ export default class SubItemsModule extends Component {
+        this.updateListViewHeight();
+        this.setState((state) => ({
+            sortClause,
+            sortOrder: state.sortClause === sortClause ? (state.sortOrder === 'ascending' ? 'descending' : 'ascending') : 'ascending',
```
This ternary operator is to complicated.

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -293,72 +431,91 @@ export default class SubItemsModule extends Component {
+        const { contentTypesMap } = this.state;
+        const notAvailableLabel = Translator.trans(/*@Desc("N/A")*/ 'content_type.not_available.label', {}, 'sub_items');
+        const contentType = contentTypesMap[item.content.ContentType._href];
+        const contentTypeName = contentType ? contentType.names.value[0]['#text'] : notAvailableLabel;
```
This will return the name in the first language. The translation will not work.

tischsoic replied:
Hmm, I took this piece of code from somewhere else, we should use `window.eZ.adminUiConfig.contentTypeNames` now?

dew326 replied:
Yes

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -557,6 +718,23 @@ export default class SubItemsModule extends Component {
+    onPageChange(page) {
+        this.updateListViewHeight();
+        this.setState(() => ({
+            activePage: page - 1,
```
Why activePage is page - 1?

tischsoic replied:
Because in the pagination the page numbers start from 1 and thus pagination button is passing `n`, not `n-1` (e.g 1 instead of 0). 
I was thinking about it and your comment convinced me that it should be changed.

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -572,23 +750,46 @@ export default class SubItemsModule extends Component {
+        const { activePage, pages } = this.state;
+        const pageData = pages[activePage];
+        const pageItems = pageData ? pageData.items : [];
+        const viewingCount = pageItems ? pageItems.length : 0;
```
hmm `pageItems` is an array (empty or populated), so `pageItems.length` is enough?

tischsoic replied:
No 😄 `pageData` contains `{ isLoading: boolean, items: Array }` and items may be null, because in `setPageLoadingState` I set the pageData to: `{ isLoading: true, items: null }`.

I'm setting it to `null`, because I have this condition:
```js
const activePageData = pages[activePage];
const showSpinner = !activePageData || !activePageData.items;

if (showSpinner) {
     return this.renderSpinner();
}
```
in `renderListView`, because when we don't have subitems in given location we want to pass empty array to `SubItemsListComponent` (as before) and `<NoItemsComponent>` is rendered by them. What do you think? Maybe I should change it so that we would render `<NoItemsComponent>` in `sub.items.component.js` based on `state.totalCount`? It may be simpler.

dew326 replied:
It seems like overcomplicated.

Correct me if I'm wrong:
1. `renderSpinner()` - renders spinner when items are loading.
2. You have flag `isLoading`
3. You render spinner when there are no items

Why not when `isLoading` is `true`?

tischsoic replied:
> Why not when isLoading is true?

Because on the first hit you don't have `pageData` for given `activePage` - it is set in `componentDidUpdate`, so on the first hit `pageData` is undefined so basically `pageData. isLoading == false` (because it is undefined) but you should treat it as `isLoading`.

dew326 replied:
It feels like something is wrong in the architecture, we have flag `isLoading` but the flag doesn't answer the question if page is loading.

sunpietro replied:
We should definitely use this flag to display the loading spinner, unless we want to mess up our code. If the `pageData` is undefined, then you have to find the answer why is it set this way.
On first render we get all data for subitems from the backend, so no initial loading is required. You have to check how the code works and understand it. Adding additional code that duplicates existing solutions is not right way of development.

tischsoic replied:
> If the pageData is undefined, then you have to find the answer why is it set this way.

I think that my explanation wasn't clear enough, when saying this:
> Because on the first hit you don't have pageData for given activePage - it is set in componentDidUpdate, so on the first hit pageData is undefined so basically pageData. isLoading == false (because it is undefined) but you should treat it as isLoading.

by `on the first hit` I was talking about "first hit" after switching to a page not yet loaded - so I don't have data for it. I trigger loading in `componentDidUpdate`. :) 

So it's like:
1. switch to a new page (not loaded yet),
2. render triggers after state change (`pageData` not set),
3. detect in `componentDidUpdate` that activePage has not been loaded yet and trigger loading (set `isLoading` flag).

sunpietro replied:
Maybe you should set base, empty state for a page without data when trying to load it?

tischsoic replied:
I will try. :)

tischsoic commented:
I've just realised that we have `activePage` in sub.items.module.js and `currentPage` in `pagination.component.js`. Which name do we want? Or maybe leave it as it is?
ping @sunpietro @dew326

sunpietro commented on Resources/translations/sub_items.en.xliff:
```diff
@@ -111,6 +101,21 @@
+        <note>key: pagination.next</note>
+      </trans-unit>
+      <trans-unit id="c6a31daba40a2cc95badf8a9c20773f144f1601b" resname="pagination.previous">
+        <source>Back</source>
```
The identifier says `previous` the label says `back`. Maybe you should consolidate the naming convention?

tischsoic replied:
I was thinking about it, I will change this.

sunpietro commented on src/modules/sub-items/components/pagination/pagination.btn.component.js:
```diff
@@ -0,0 +1,38 @@
+        }
+    };
+
+    let className = `page-item ${additionalClasses}`;
```
The CSS class name is incosistent with the convention exisiting in this repo.

tischsoic replied:
You mean `page-item`? AFAIK it's Bootstrap class, do you want me to duplicate it?

sunpietro replied:
If it's from Bootstrap then it's fine.

tischsoic replied:
OK 👍

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,126 @@
+                pages.push(i);
+                wereDots = false;
+            } else if (!wereDots) {
+                pages.push('...');
```
I would convert dots into a kind of static variable that could be reused in several places.

```suggestion
                const DOTS = '...';
                pages.push(DOTS);
```

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,126 @@
+
+        return (
+            <div className="c-pagination row justify-content-center align-items-center">
+                <ul className="pagination ez-pagination">
```
Incorrect CSS class name convention

tischsoic replied:
The same as above.

sunpietro commented on src/modules/sub-items/components/table-view/table.view.component.js:
```diff
@@ -18,140 +20,42 @@ export default class TableViewComponent extends Component {
-        } else {
-            this.setState(() => ({ items }));
-        }
+        onSortChange('ContentName');
```
I would make it a constant like `const KEY_CONTENT_NAME = 'ContentName';`. Then in the:

```
const SORTKEY_MAP = {
    [KEY_CONTENT_NAME]: 'name',
    DateModified: 'date',
    LocationPriority: 'priority',
};
```

sunpietro replied:
The similar rule applies to other places.

sunpietro commented on src/modules/sub-items/components/viewing-message/viewing.message.component.js:
```diff
@@ -0,0 +1,30 @@
+    viewingCount: PropTypes.number.isRequired,
+};
+
+ViewingMessageComponent.defaultProps = {};
```
It can be removed.

sunpietro commented on src/modules/sub-items/components/table-view/table.view.component.js:
```diff
@@ -18,140 +20,42 @@ export default class TableViewComponent extends Component {
-            this.setState(() => ({ items }));
-        }
+    sortByName() {
+        const { onSortChange } = this.props;
```
Where is this method/function defined?

tischsoic replied:
in `sub.items.module.js`, it is past via props here: https://github.com/ezsystems/ezplatform-admin-ui-modules/pull/112/files#diff-bebc1f75bde44101c137d4bc83684925R872

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +304,57 @@ export default class SubItemsModule extends Component {
+        this.updateListViewHeight();
+        this.setState((state) => ({
+            sortClause,
+            sortOrder: state.sortClause === sortClause ? this.getOppositeSortOrder(state.sortOrder) : this.getDefaultSortOrder(),
```
I would expose it as a class method: `getSortOrder`: `state.sortClause === sortClause ? this.getOppositeSortOrder(state.sortOrder) : this.getDefaultSortOrder()`

What do you think about this approach?

tischsoic replied:
This was my first approach but `getSortOrder` is not very descriptive and you have to pass `state.sortClause`, `state.sortOrder` and `sortClause` to this method which for me obscures what it does.
But if you prefer `getSortOrder` I can change it. :)

sunpietro replied:
I think, `getSortOrder` makes sense.

sunpietro commented on src/modules/sub-items/components/pagination/pagination.btn.component.js:
```diff
@@ -0,0 +1,38 @@
+
+    let className = `page-item ${additionalClasses}`;
+
+    className = disabled ? `${className} disabled` : className;
```
Is the `disabled` class coming from Bootstrap? Why not to add `disabled` attribute on the button?

tischsoic replied:
Styling does not work if I switch to attribute.

sunpietro replied:
Then I would create a generic selector: `[disabled] { ... }` in `vendor/ezsystems/ezplatform-admin-ui/src/bundle/Resources/public/scss/_general.scss` and I'd add the properties required to display a disabled element the proper way. What do you think about it @dew326 ?

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,126 @@
+
+import PaginationBtnComponent from './pagination.btn.component';
+
+export default class PaginationComponent extends PureComponent {
```
It should be a functional component instead.

tischsoic replied:
I decided to use class component because with that many functions I think it is more readable and clear. Do you want me to change it to functional component and change each method to arrow function within it?

sunpietro replied:
The functions inside the class can be easily converted to a _standard_ arrow functions outside of functional component. They will take props as arguments and return a value.
I believe it's better not to define such functions inside the functional component as they're not very specific.

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    renderNext() {
+        const { currentPage, onPageChange } = this.props;
```
I would rename `currentPage` to `currentPageIndex`

tischsoic replied:
Mhm, I like this idea, you want to add `Index` to suggest that numeration in `currentPage` variable starts from 0? :)

sunpietro replied:
That would make sense. And it should be `activePage`, right?

tischsoic replied:
Right 👍

tischsoic replied:
I changed this but it might be a little bit inconsistent to have same value under `currentPage` and `currentPageIndex.`. 🤔  Nonetheless, I will keep it as it is now, after correction.

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,126 @@
+            }
+
+            const isCurrentPage = page === currentPage + 1;
+            const additionalClasses = isCurrentPage ? 'active' : '';
```
Is `active` CSS class coming from Bootstrap?

tischsoic replied:
Yes, https://getbootstrap.com/docs/4.1/components/pagination/#disabled-and-active-states

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,126 @@
+        const pages = this.computePages();
+
+        return (
+            <div className="c-pagination row justify-content-center align-items-center">
```
I bet, the `row` CSS class should not be here.

tischsoic replied:
This is how it is styled in other places, we use `row` css class.

sunpietro replied:
Shouldn't the `row` class be applied to the nested element instead: `ul`? Because it will be the row actually.

tischsoic replied:
It can be moved to `ul`, I will do this.

tischsoic replied:
I moved CSS classes to `ul` element and removed `div`.

sunpietro commented:
> I've just realised that we have `activePage` in sub.items.module.js and `currentPage` in `pagination.component.js`. Which name do we want? Or maybe leave it as it is?

I would stick with `activePage`

tischsoic commented:
ping @dew326 @sunpietro

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -293,72 +447,96 @@ export default class SubItemsModule extends Component {
+        const notAvailableLabel = Translator.trans(/*@Desc("N/A")*/ 'content_type.not_available.label', {}, 'sub_items');
+        const contentType = contentTypesMap[item.content.ContentType._href];
+        const contentTypeIdentifier = contentType.identifier;
+        const contentTypeName = window.eZ.adminUiConfig.contentTypeNames[contentTypeIdentifier] || notAvailableLabel;
```
If you have `contentTypeIdentifier` it will always have the name in the object `contentTypeNames`. I mean `|| notAvailableLabel` will never be shown. 
The only one what you may not have is content type in `contentTypesMap` but it will throw an exception earlier when you try to read property `identifier`.

tischsoic replied:
Fix amended to last commit.

sunpietro commented on src/modules/sub-items/components/pagination/pagination.btn.component.js:
```diff
@@ -0,0 +1,38 @@
+    className = disabled ? `${className} disabled` : className;
+
+    return (
+        <li className={className}>
```
I was thinking about this situation when a component uses only Bootstrap CSS classes for styling. I've came up with an idea that it should have at least one CSS class that tell us that this is a `PaginationButtonComponent`. I'd add a CSS class: `c-pagination-button` to the `li` element.

sunpietro commented on src/modules/sub-items/components/pagination/pagination.btn.component.js:
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const PaginationBtnComponent = ({ label, disabled, additionalClasses, onPageChange, page }) => {
```
`PaginationButtonComponent`

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,101 @@
+    const isFirstPage = activePageIndex === 0;
+    const isLastPage = activePageIndex === pagesNumber - 1;
+    const pages = computePages({ proximity, activePageIndex, pagesNumber });
+    const pagesElements = pages.map((page, index) => {
```
What does `page`variable contain?

tischsoic replied:
One value of the array: `[ 1, "...", 3, 4, 5, 6, 7, "...", 10 ]`.

sunpietro replied:
`paginationButtons`

tischsoic replied:
I'm not sure, you want to rename `pagesElements` to `paginationButtons` or `page` to `paginationButtons`?

sunpietro replied:
I'm sorry for not being clear enough here. I thought about converting `pagesElements` to `paginationButtons` becuase that's what this variable contains.

tischsoic replied:
Right, it makes sense! :)

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,101 @@
+    });
+
+    return (
+        <ul className="pagination row justify-content-center ez-pagination">
```
I'd add a CSS class: `c-pagination`

sunpietro replied:
Do you need `ez-pagination` class?

tischsoic replied:
It seems that we can remove `ez-pagination` when it comes to solely styling.

sunpietro commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,101 @@
+ *
+ * @returns {Array}
+ */
+const computePages = ({ proximity, activePageIndex, pagesNumber }) => {
```
Please, place it before component definition.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -60,9 +74,26 @@ export default class SubItemsModule extends Component {
+        const pageDoesNotExist = activePage > numberOfPages - 1 && activePage !== 0;
+
+        if (pageDoesNotExist) {
+            this.setState({
```
This is risky, as it might lead to infinite loop. We have to be really careful when updating state in the lifecycle method that is run right after a state is updated.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -141,51 +199,54 @@ export default class SubItemsModule extends Component {
+                const totalCount = response.View.Result.count;
+
+                if (oldTotalCount !== totalCount) {
+                    this.discardAllPagesOtherThan(page);
```
This is strange approach for naming methods: `discardAllPagesOtherThan()`. I would rename it to `discardPages(<array-of-excluded-pages-from-discard>)`.
What is a result of discard action?

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +304,61 @@ export default class SubItemsModule extends Component {
+        }));
+    }
+
+    discardCurrentPage() {
```
`discardCurrentPageData`

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +304,61 @@ export default class SubItemsModule extends Component {
+        }));
+    }
+
+    discardAllPages() {
```
`discardAllPagesData`

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +304,61 @@ export default class SubItemsModule extends Component {
+        }));
+    }
+
+    discardAllPagesOtherThan(page) {
```
This is redundant. This case can be covered in `discardAllPagesData` method

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -259,27 +378,62 @@ export default class SubItemsModule extends Component {
+        let itemIndex = null;
 
-        item.location = location;
+        for (const [page, pageData] of Object.entries(pages)) {
```
Why not `Array.find`?

tischsoic replied:
I think I have chosen this because `for of` with `break;` seemed for me more clear than `Object.entries(pages).find(callback)` with callback returning `true` after finding given page.

tischsoic replied:
For me both ways are OK. Do you want me to convert this to `Array.find()`?

sunpietro replied:
Yes, please

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -293,72 +447,95 @@ export default class SubItemsModule extends Component {
+     *
+     * @returns {Object} dataToCache
+     */
+    getItemDataToCache(item) {
```
I don't get the idea of caching here. Can you elaborate more on this topic? What is the flow?

tischsoic replied:
You cache item data during a selection of a page item so that you have item data even if page data were discarded.

tischsoic replied:
When it comes to an example of flow: 
1. you open the site, Sub-items 1st page has been loaded
2. you sort items by name *ascending*
3. 1st page reloads
4. you select 1st item on the 1st page
5. you sort items by name again but *descending*
6. 1st page reloads (now selected item is on another page)
7. you select another 1st item on 1st page
8. you perform bulk delete in which both items should be moved to trash

sunpietro replied:
So technically it's not a cache. It's selected items store.
When I see cache, then it means that you can use it get full component state without a need to send requests to backend.

tischsoic commented:
ping @sunpietro @dew326 
New changes in last commit.
I applied fixes according to PR review. I added new variable `pagesDuringLoading` which stores set of pages during loading so that pages data in state are only items - no loading information. We show spinner when `pageItems` are `undefined`. I moved `<NoItemsComponent />` higher to simplify logic. +other small changes

tischsoic commented:
ping @sunpietro @dew326

dew326 commented on src/modules/sub-items/components/pagination/pagination.button.component.js:
```diff
@@ -0,0 +1,38 @@
+    label: PropTypes.string.isRequired,
+    disabled: PropTypes.bool.isRequired,
+    onPageChange: PropTypes.func,
+    page: PropTypes.number,
```
`page` seems to be required and `disabled` seems to be not required.

tischsoic replied:
In previous implementation `page` was not required, we can make it required. When it comes to  `disabled` I think it may stay as required - we have it in `defaultProps`.

dew326 replied:
It's not about what we can do or not, it's about which props have to be provided to make the component work properly.

To make component work you have to pass the `page` - so this is required
If you will not provide the `disabled` flag the default value will be used - so this is not required.

tischsoic replied:
Oh, now I checked it more carefully, and we have:
```js
if (page === DOTS) {
     return <PaginationButtonComponent key={index} label={DOTS} disabled={true} />;
}
```
so `page` prop cannot be required.

dew326 replied:
Ok, but `disabled` is not required.

tischsoic replied:
Correct, I was during fixing this.

dew326 commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,101 @@
+    const previousPage = activePageIndex - 1;
+    const nextPage = activePageIndex + 1;
+    const isFirstPage = activePageIndex === 0;
+    const isLastPage = activePageIndex === pagesNumber - 1;
```
Maybe better would be `const isLastPage = activePageIndex + 1 === pagesNumber` because `pagesNumber - 1` suggests before last page.

dew326 commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -60,9 +73,27 @@ export default class SubItemsModule extends Component {
+        const { activePage, pages, totalCount } = this.state;
+        const { limit: itemsPerPage } = this.props;
+        const numberOfPages = Math.ceil(totalCount / itemsPerPage);
+        const pageDoesNotExist = activePage > numberOfPages - 1 && activePage !== 0;
```
Why it is possible to have `activePage` bigger than `pages count`?

tischsoic replied:
If you are on the last page and delete all items from the last page this last page disappears (no longer exists). This mechanism switches the active page to the last one.

sunpietro commented on src/modules/sub-items/components/table-view/table.view.component.js:
```diff
@@ -228,9 +117,11 @@ export default class TableViewComponent extends Component {
+        const { sortClause, sortOrder } = this.props;
+
+        if (sortClause) {
+            const headSortClass = sortOrder === 'ascending' ? `${headClass}--sort-asc` : `${headClass}--sort-desc`;
```
`headSortOrderClass`

sunpietro commented on src/modules/sub-items/components/table-view/table.view.component.js:
```diff
@@ -292,4 +183,9 @@ TableViewComponent.propTypes = {
+    sortOrder: PropTypes.string.isRequired,
 };
+
+TableViewComponent.defaultProps = {};
```
Can be removed.

sunpietro commented on src/modules/sub-items/components/viewing-message/viewing.message.component.js:
```diff
@@ -0,0 +1,28 @@
+
+import './css/viewing.message.component.css';
+
+const ViewingMessageComponent = ({ totalCount, viewingCount }) => {
```
Isn't `PaginationInfoComponent` name better than this, is it?

tischsoic replied:
It may be. 😝

dew326 commented on src/modules/sub-items/components/pagination/pagination.component.js:
```diff
@@ -0,0 +1,101 @@
+    const pages = computePages({ proximity, activePageIndex, pagesNumber });
+    const pagesElements = pages.map((page, index) => {
+        if (page === DOTS) {
+            return <PaginationButtonComponent key={index} label={DOTS} disabled={true} />;
```
Using `index` and `page` as key may cause conflicts.
In case: `[ 1, "...", 3, 4, 5, 6, 7, "...", 10 ]`
First iteration: `page === 1`
Second iteration: `index === 1`
In two items key is `1`

Maybe we should use a combination of page and index? like `page-index`?

tischsoic replied:
Nice catch! 👍 
When we have dots, under page variable we have: `...`. I think we should be fine with a key like: `"dots-${index}"`.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -21,35 +25,44 @@ export default class SubItemsModule extends Component {
+        this._refListViewWrapper = React.createRef();
         this.bulkDeleteModalContainer = null;
         this.udwContainer = null;
+        this.pagesDuringLoading = new Set();
```
This property names still looks enigmatic to me. After getting back to your code, for review, I don't know what is it for.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -21,35 +25,44 @@ export default class SubItemsModule extends Component {
-            items: props.items,
-            selectedLocationsIds: new Set(),
+            pages: {
+                '0': props.items,
```
`0: props.items` looks good as well.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -21,35 +25,44 @@ export default class SubItemsModule extends Component {
+            pages: {
+                '0': props.items,
+            },
+            selectedItemsCachedData: new Map(),
```
I still believe that caching is not needed and in practice it's something different than the variable name says to us.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -60,9 +73,27 @@ export default class SubItemsModule extends Component {
+    componentDidUpdate() {
+        const { activePage, pages, totalCount } = this.state;
+        const { limit: itemsPerPage } = this.props;
+        const numberOfPages = Math.ceil(totalCount / itemsPerPage);
```
`pagesCount`

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -74,14 +105,23 @@ export default class SubItemsModule extends Component {
-     */
-    handleLoadMore() {
-        this.setState((state) => ({ offset: state.offset + this.props.limit }));
+    extractFirstSortClauseData(sortClauses) {
```
Maybe `getDefaultSortClause`?

tischsoic replied:
I prefer `extractFirstSortClauseData` over `getDefaultSortClause` because although maybe it is a little bit uglier, I know what it does inside. If you remove context from it, function name still makes sense. This is how I reason about this.

sunpietro replied:
I prefer naming methods/functions in the way that I know what it does and what's its purpose. You need this method to get the sort clause when initializing the component, hence why I suggested the name `getDefaultSortClause`. Because when you're initializing something, you'll probably want something to have a default value, like in your code.
If the method was used in many different places, in many different contexts then your reasoning might be correct. Otherwise I prefer my proposal.

dew326 replied:
I agree with @sunpietro. The method returns the default sort clause.
If for some reason in initial state you would need to have last sort clause, you would like to introduce the new method `extractLastSortClause`?

tischsoic replied:
For me both trains of thought are OK, and since we have 2:1 I will change this to `getDefaultSortClause`. :)

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -90,32 +130,40 @@ export default class SubItemsModule extends Component {
     /**
      * Loads location data
      *
+     * @param {Number} page
```
Maybe `pageNumber` or `pageIndex`?

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -124,11 +172,12 @@ export default class SubItemsModule extends Component {
      * Loads content items from location
      *
      * @method loadContentItems
+     * @param {Number} page
```
Maybe `pageNumber` or `pageIndex`?

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -141,51 +190,54 @@ export default class SubItemsModule extends Component {
+                    page,
+                    locations,
+                    totalCount,
+                    contents: contentInfo.View.Result.searchHits.searchHit.map((searchHit) => searchHit.value.Content),
```
`content` is singular only, hence you should name it `contentItems`

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +297,42 @@ export default class SubItemsModule extends Component {
+        }));
+    }
+
+    discardPagesDataOtherThan(pages) {
```
When is this method invoked? Because the method name is a bit messy.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +297,42 @@ export default class SubItemsModule extends Component {
+    }
+
+    getDefaultSortOrder() {
+        return 'ascending';
```
Why not setting it as a const? `const DEFAULT_SORT_ORDER = 'ascending';`

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -240,6 +297,42 @@ export default class SubItemsModule extends Component {
+    }
+
+    getOppositeSortOrder(sortOrder) {
+        return sortOrder === 'ascending' ? 'descending' : 'ascending';
```
`ascending` and `descending` can be set as constant variables outside the class definition.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -259,27 +352,57 @@ export default class SubItemsModule extends Component {
-     * @param {Object} state
-     * @returns {Object}
-     * @memberof SubItemsModule
+     * @returns {{itemPage: number, itemIndex: number}}
```
`pageNumber` and `pageIndex`, maybe?

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -293,72 +416,95 @@ export default class SubItemsModule extends Component {
+        return {
+            contentTypeName,
+            name: item.content.Name,
+            location: {
```
Why not storing the full `item` object instead?

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -293,72 +416,95 @@ export default class SubItemsModule extends Component {
-        const contentItemsIds = items.map((item) => item.location.id);
-        const contentItemsIdsSet = new Set(contentItemsIds);
+        if (select) {
+            const itemsCachedData = new Map(pageItems.map((item) => [item.location.id, this.getItemDataToCache(item)]));
```
Do you really need to create a `Map` instance?

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -293,72 +416,95 @@ export default class SubItemsModule extends Component {
+        this.setSelectedItemsState(new Map());
+    }
+
+    setSelectedItemsState(selectedItemsCachedData) {
```
Why not calling `setState` directly. It's also one function with one param.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -605,20 +794,81 @@ export default class SubItemsModule extends Component {
+    renderNoItems() {
+        const { totalCount } = this.state;
+
+        if (totalCount) {
```
`if (this.state.totalCount) {` is just enough

sunpietro reviewed (changes requested):
The concept of cache for selected items doesn't look very well. Especially, that it doesn't work like cache. It's more like a store.
Other things are minor compared to the cache.

tischsoic committed:
Small rewrite of the pagination - now each time the page is changed it is reloaded

tischsoic commented:
I applied corrections - ping @sunpietro  @dew326

sunpietro commented on src/modules/sub-items/components/pagination-info/pagination.info.component.js:
```diff
@@ -0,0 +1,28 @@
+    );
+
+    return (
+        <div className="c-pagination-info">
```
I think this `div` is not needed.

sunpietro commented on src/modules/sub-items/components/pagination-info/pagination.info.component.js:
```diff
@@ -0,0 +1,28 @@
+
+import './css/pagination.info.component.css';
+
+const PaginationInfoComponent = ({ totalCount, viewingCount }) => {
```
I'm not sure we even need this component as this will be only one `div` if you follow the comment below.

tischsoic replied:
We can make it with one `div` - I will delete this component and move code to sub.items.module.js.

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -259,27 +323,34 @@ export default class SubItemsModule extends Component {
-        const items = state.items;
-        const item = items.find((element) => element.location.id === location.id);
+    updateItemLocation(location) {
+        this.setState((state) => {
```
`this.setState(({ activePageItems }) => {`

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -605,20 +708,78 @@ export default class SubItemsModule extends Component {
+
+        return (
+            <div style={style}>
+                <div className="m-sub-items__list-spinner-wrapper">
```
`m-sub-items__spinner-wrapper` maybe?

tischsoic replied:
`spinner-wrapper` is better 👍

sunpietro commented on src/modules/sub-items/sub.items.module.js:
```diff
@@ -605,20 +708,78 @@ export default class SubItemsModule extends Component {
+        return (
+            <div style={style}>
+                <div className="m-sub-items__list-spinner-wrapper">
+                    <Icon name="spinner" extraClasses="m-sub-items__list-spinner ez-spin ez-icon-x2 ez-icon-spinner" />
```
`m-sub-items__spinner`

tischsoic committed:
fix - ping @sunpietro

sunpietro reviewed (approved):
Seems to be it. Let's see what @dew326 says about it

SylvainGuittard commented:
🎉 🎉 🎉 
Congrats @tischsoic and thanks @sunpietro @dew326 for the reviews!

barbaragr reviewed (approved):
Works fine 

🐴

mnocon commented:
When merging: please merge https://github.com/ezsystems/ezplatform-admin-ui/pull/710 as well (it adjusts tests)
